include/deal.II/lac/precondition.h

```diff
@@ -3157,44 +3148,41 @@ namespace internal
 
       if (iteration_index == 0)
         {
-          const auto solution_old_ptr = solution_old.begin();
-
           // compute t = P^{-1} * (b)
           preconditioner.vmult(solution_old, rhs);
 
           // compute x^{n+1} = f_2 * t
+          const auto solution_old_ptr = solution_old.begin();
```

Comment:
This was the place where step-59 ran into a 'use-after-free' problem: Inside `preconditioner.vmult()`, we would create a new vector with new data array, thus invalidating the `solution_old_ptr` that we captured before. Now the question is of course whether we would like to allow changes in the storage (and several other places I touched above do still suffer from this problem), but I think here the case is clear-cut that we should define the shortcut near the place of use.